File: node/src/memory/in-memory-store.ts
/**
 * `InMemoryDreamStore` — a lightweight `DreamStore` implementation backed by per-agent `Map`s.
 *
 * Originally lived as `MockDreamStore` in the SDK's test helpers; promoted here so benchmarks,
 * examples, and downstream consumers can use it without copying the boilerplate.
 *
 * Use cases:
 *   - Benchmark A/B variants where memory is on/off (preload via constructor).
 *   - Unit tests that exercise session extraction or the `memory_query` path without disk I/O.
 *   - Local development / CI where a persistent memory store isn't needed.
 *
 * Search is a deterministic local reference implementation: distinct lexical overlap first,
 * `metadata.updated_at` recency second, and insertion order as the final stable tie-breaker.
 * Non-empty queries never return unrelated entries. For semantic search, plug in a real store.
 */
import type {
  DreamStore,
  MemoryQuery,
  MemoryRecall,
  MemoryRecord,
  SessionData,
} from "./protocols.js"
import { rankMemories } from "./ranking.js"

export class InMemoryDreamStore implements DreamStore {
  private memories = new Map<string, MemoryRecord[]>()
  /** Sessions persisted via `saveSession`; exposed for test assertions. */
  readonly savedSessions: SessionData[] = []

  /**
   * @param initialMemories Optional seed memories applied to every agent that asks for memories
   *                        for the first time. Useful for benchmark scenarios that preload a fact.
   */
  constructor(private readonly initialMemories: MemoryRecord[] = []) {}

  private recordsFor(agentId: string): MemoryRecord[] {
    if (this.memories.has(agentId)) return this.memories.get(agentId)!
    if (this.initialMemories.length > 0) {
      this.memories.set(agentId, [...this.initialMemories])
      return this.memories.get(agentId)!
    }
    return []
  }

  async upsert(agentId: string, incoming: MemoryRecord): Promise<void> {
    const kept = [...this.recordsFor(agentId)]
    const index = kept.findIndex(record =>
        record.scope.tenant_id === incoming.scope.tenant_id
        && record.scope.namespace === incoming.scope.namespace
        && record.kind === incoming.kind
        && record.name === incoming.name,
    )
    if (index >= 0) kept[index] = incoming
    else kept.push(incoming)
    this.memories.set(agentId, kept)
  }

  async search(agentId: string, query: MemoryQuery): Promise<MemoryRecall[]> {
    const all = this.recordsFor(agentId)
    const candidates = all.filter(record =>
      record.scope.tenant_id === query.scope.tenant_id
      && record.scope.namespace === query.scope.namespace
      && (query.kinds.length === 0 || query.kinds.includes(record.kind))
      && (query.min_score === undefined || record.confidence >= query.min_score),
    )
    const selected = rankMemories(query.query, candidates.map((record, insertionIndex) => {
      return {
        value: record,
        searchableText: `${record.name} ${record.description} ${record.content}`,
        updatedAt: Number.isFinite(record.updated_at) ? record.updated_at : 0,
        insertionIndex,
      }
    }), query.top_k)
    return selected.map(record => ({
      record,
      score: Math.max(0, Math.min(1, record.confidence)),
      why: "deterministic lexical relevance with recency tie-breaking",
    }))
  }

  async saveSession(data: SessionData): Promise<void> {
    this.savedSessions.push(data)
  }
}

File: python/deepstrike/memory/in_memory_store.py
"""`InMemoryDreamStore` — lightweight `DreamStore` backed by per-agent dicts.

Python port of node/src/memory/in-memory-store.ts. Use for benchmarks, unit tests,
and local development where persistent memory isn't needed.
"""
from __future__ import annotations

from typing import Iterable

from deepstrike.memory.protocols import DreamStore, MemoryQuery, MemoryRecall, MemoryRecord, SessionData
from deepstrike.memory.ranking import rank_memories


class InMemoryDreamStore(DreamStore):
    def __init__(self, initial_memories: Iterable[MemoryRecord] | None = None) -> None:
        self._memories: dict[str, list[MemoryRecord]] = {}
        self._initial: list[MemoryRecord] = list(initial_memories or [])
        #: Sessions persisted via save_session — exposed for test assertions.
        self.saved_sessions: list[SessionData] = []

    def _records_for(self, agent_id: str) -> list[MemoryRecord]:
        if agent_id in self._memories:
            return list(self._memories[agent_id])
        if self._initial:
            self._memories[agent_id] = list(self._initial)
            return list(self._memories[agent_id])
        return []

    async def upsert(self, agent_id: str, incoming: MemoryRecord) -> None:
        kept = self._records_for(agent_id)
        index = next((i for i, record in enumerate(kept) if record.scope == incoming.scope and record.kind == incoming.kind and record.name == incoming.name), None)
        if index is None:
            kept.append(incoming)
        else:
            kept[index] = incoming
        self._memories[agent_id] = kept

    async def search(self, agent_id: str, query: MemoryQuery) -> list[MemoryRecall]:
        candidates = [record for record in self._records_for(agent_id)
                      if record.scope == query.scope
                      and (not query.kinds or record.kind in query.kinds)
                      and (query.min_score is None or record.confidence >= query.min_score)]
        selected = rank_memories(
            query.query, candidates, query.top_k,
            searchable_text=lambda record: f"{record.name} {record.description} {record.content}",
            updated_at=lambda record: record.updated_at,
        )
        return [MemoryRecall(record=record, score=max(0.0, min(1.0, record.confidence)), why="deterministic lexical relevance with recency tie-breaking") for record in selected]

    async def save_session(self, data: SessionData) -> None:
        self.saved_sessions.append(data)
